Core Concept week 1 day 5 update (#39)
microsoft/win-student-devs: docs/30DaysOfPWA/core-concepts/05.md

**Author: Nitya Narasimhan [@nitya](https://twitter.com/nitya)**

Welcome to **Day 5** of #30DaysOfPWA! Want to learn more about this project? Check out our [Kickoff](../kickoff.md) post to get more details on the content roadmap and contributors. Now, let's dive in!

![Day 5: Let's Talk Service Workers!](_media/day-05.jpg)

|  | What you will learn today |
|:--- |:---|
| _Offline PWA_ | What does offline mean? What should the PWA experience be? |
| _Detection_ | How does PWA know about network changes? |
| _Storage_ | What storage options can service workers use? |
| _Caching_ | What caching strategies can PWAs employ? |
| _Exercise_ | Explore offline experience in sample PWA. |
| _Related_ | [Week 3: Developer Tools](../dev-tools/) |

---

## Let's Recap

**What we learned so far:**

 * PWAs are web apps that use _progressive enhancement_ to scale experiences up and down to suit device and browser capabilities. On more capable devices, they can be indistinguishable from natively installed apps.
 * PWAs use _open web technologies_ with core building blocks being HTTPS (make it secure), Service Workers (make it reliable) and Web App Manifest (make it installble).

>  * _Service Workers_ are web workers that run in the background, executing time-consuming, asynchronous tasks without impacting page performance and responsiveness to users.

Author: How do you suggest we modify it here?

>  * _Service Workers_ are web workers that run in the background, executing time-consuming, asynchronous tasks without impacting page performance and responsiveness to users.

Reviewer: "Web worker" could be a link to MDN if they want to read more about regular workers.

>  * _Service Workers_ are web workers that run in the background, executing time-consuming, asynchronous tasks without impacting page performance and responsiveness to users.

Reviewer 2: What if we removed the sections below this and replaced the above line with a single line (linking to web workers on MDN as Aaron suggested)? 

Service Workers are a type of web worker that acts as a network proxy (intercepting fetch requests), manages cache storage intelligently (for offline operation) and handles async events like push notifications (to support re-engageable behaviors).

>  * _Service Workers_ are web workers that run in the background, executing time-consuming, asynchronous tasks without impacting page performance and responsiveness to users.

Reviewer 2: ```suggestion
 * Service Workers are a type of web worker that acts as a network proxy (intercepting fetch requests), manages cache storage intelligently (for offline operation) and handles async events like push notifications (to support re-engageable behaviors).
```

> 
> **Yesterday, we learned about Service Workers:**
> 
> * They are **Network proxies**. They _intercept and modify_ app requests for network resources, making HTTPS usage mandatory. Their ability to modify network messages means they need HTTPS to prevent _others_ from using that ability to hijack client-server communications.
> * They are **Cache managers.** They _precache assets_ and manage cache storage using pre-defined strategies. This means the PWA can operate offline (using pre-cached offline pages and responses) and be more performant (e.g., pre-caching assets with longer lifetimes).
> * They are **Async Event handlers.** They _receive push notifications_ even if the app is inactive - and can respond by triggering device-familiar alerts that can _re-engage the user_  to interact with the app, just like in other native experiences.
> 
> **And we learned about the Service Worker Lifecycle**:
> 
> Service Workers handle fetch events for app resources (paths) within their "scope". To minimize conflict with existing service workers (wrapping up prior requests), they are brought to life in three phases, with the install-activate transition waiting on existing service workers to exit.
> 
>  * _registration_ - service worker is downloaded and initialized.
>  * _installation_ - service worker settles in, gets first event.
>  * _activation_ - service worker is greenlit for active duty.
> 
> Using our "service worker as Chief Operations Officer" analogy - it's like the startup hired a new COO (registration), gave them their badge (install event), then waited till the prior COO cleared the office before handing over the key (activate event). The badge allows the COO to bring in things they need (pre-cache) but it's the key that gives them the power to handle calls (functional events) and redecorate (e.g., clear out previous version's cache).
> 

Reviewer 2: ```suggestion

```

Here's the relevant snippet from our sample PWA.

```javascript

// We have a cache-first strategy, 
// where we look for resources in the cache first
// and only on the network if this fails.
self.addEventListener('fetch', event => {
    event.respondWith((async () => {
        const cache = await caches.open(CACHE_NAME);

        // Try the cache first.
        const cachedResponse = await cache.match(event.request);
        if (cachedResponse !== undefined) {
            // Cache hit, let's send the cached resource.
            return cachedResponse;
        } else {
            // Nothing in cache, let's go to the network.

            // ...... truncated ....
        }
    }
}
```

---

## Detecting Network Changes

The above examples showcased scenarios where the caching policy was independent of network status (i.e., always checks cache first). But what if you wanted to condition your strategy on the current status of the network?

> The [`navigator.onLine`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/onLine) property returns a boolean (true/false) value reflecting the online status of the browser. Different browsers might implement this differently - so you may want to understand nuances to avoid false positives and negatives. The property should send update events if that status changes - and you can listen for those events at the `window` level.

Reviewer 2: YES - good suggestions all and I can fix this now. Thank you so much @aarongustafson @zateutsch

```javascript
window.addEventListener("online",  function(){
    console.log("You are online!");
});
window.addEventListener("offline", function(){
    console.log("Oh no, you lost your network connection.");
});

> ```

Reviewer 2: ```suggestion
```

So how can this event listener notify the Service Worker of that network change? Service Workers and their clients can communicate using the [postMessage](https://developer.mozilla.org/en-US/docs/Web/API/Client/postMessage) API.  On the client side,  this involves sending a message with relevant data:

```javascript
navigator.serviceWorker.controller.postMessage({ 
   type: `IS_OFFLINE`
   // add more properties if needed
});
```

On the service worker side, it listens for "message" events and unpacks the data to take further action:
```javascript
self.addEventListener('message', (event) => {
  if (event.data && event.data.type === 'IS_OFFLINE') {
    // take relevant actions
  }
});
```

You can also use the [navigator.connection](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/connection) property to access information about network quality. This includes:
 * _navigator.connection.downlink_ to get effective bandwidth estimate (if available)
 * _navigator.connection.saveData_ if user activated the "save data" option in their browser

```

Author: I updated this in commit [32b9de9](https://github.com/microsoft/win-student-devs/pull/39/commits/32b9de972c333adb835ceebcc7b87f09e941a7cb)

 * *Service Worker API* | This [MDN resource](https://aka.ms/learn-PWA/30Days-1.5/developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) has good coverage of Service Worker interfaces, use cases and related APIs.
 * *Storage Options* | Learn more about [Cache Storage API](https://aka.ms/learn-PWA/30Days-1.5/developer.mozilla.org/en-US/docs/Web/API/CacheStorage), [Cache API](https://aka.ms/learn-PWA/30Days-1.5/developer.mozilla.org/en-US/docs/Web/API/Cache) and [IndexedDB API](https://aka.ms/learn-PWA/30Days-1.5/developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API) - to understand what types of data each supports, and how they can be used by service workers.
 * *Workbox* | These [Google-developed libraries](https://aka.ms/learn-PWA/30Days-1.5/developers.google.com/web/tools/workbox) power production-ready service workers and outline [common recipes](https://aka.ms/learn-PWA/30Days-1.5/developers.google.com/web/tools/workbox/guides/common-recipes) for caching strategies you can use in your PWA.

> Look out for content in Week 3 (Developer Tools) and Week 4 (Platforms & Practices) that may explore some of these topics in more detail.

Reviewer 3: Maybe just say the following weeks because we talk about caching & push in week 2.

> Look out for content in Week 3 (Developer Tools) and Week 4 (Platforms & Practices) that may explore some of these topics in more detail.

Author: ```suggestion
Look out for content in the following weeks that may explore some of these topics in more detail.
```
